tooling/nargo/src/ops/foreign_calls.rs

```diff
@@ -76,10 +126,92 @@ impl ForeignCall {
                     ],
                 })
             }
-            None => panic!("unexpected foreign call {foreign_call_name:?}"),
+            Some(ForeignCall::CreateMock) => {
+                let mock_oracle_name = Self::parse_string(&foreign_call.inputs[0]);
+                assert!(ForeignCall::lookup(&mock_oracle_name).is_none());
+                let id = self.last_mock_id;
+                self.mocked_responses.push(MockedCall::new(id, mock_oracle_name));
+                self.last_mock_id += 1;
+
+                Ok(ForeignCallResult { values: vec![Value::from(id).into()] })
+            }
+            Some(ForeignCall::SetMockParams) => {
+                let (id, params) = Self::extract_mock_id(&foreign_call.inputs)?;
+                self.find_mock_by_id(id)
+                    .unwrap_or_else(|| panic!("Unknown mock id {}", id))
+                    .params = Some(params.to_vec());
+
+                Ok(ForeignCallResult { values: vec![] })
+            }
+            Some(ForeignCall::SetMockReturns) => {
+                let (id, params) = Self::extract_mock_id(&foreign_call.inputs)?;
+                self.find_mock_by_id(id)
+                    .unwrap_or_else(|| panic!("Unknown mock id {}", id))
+                    .result = ForeignCallResult { values: params.to_vec() };
+
+                Ok(ForeignCallResult { values: vec![] })
+            }
+            Some(ForeignCall::SetMockTimes) => {
+                let (id, params) = Self::extract_mock_id(&foreign_call.inputs)?;
+                let times = params[0]
+                    .unwrap_value()
+                    .to_field()
+                    .try_to_u64()
+                    .expect("Invalid bit size of times");
+
+                self.find_mock_by_id(id)
+                    .unwrap_or_else(|| panic!("Unknown mock id {}", id))
```

Comment:
These panics could be turned into errors, but this should be rare as it is an internal compiler error so this is non-blocked if we need to get this in. `ForeignCallError` was moved into `noirc_printable_type` so we probably could do a general error refactor in a separate PR.

Reply:
Yeah it felt pretty weird to add these mocking specific errors in noirc_printable_type